### DISCUSSION TITLE
1D Diffusion Equation: Boundary Condition

### DISCUSSION THREAD

**Discussion by beantsja** [usnistgov/fipy#999] [Q&A] [answered]
Solving a 1D diffusion equation:

<img width="136" alt="image" src="https://github.com/usnistgov/fipy/assets/155326827/77340d2f-1719-4d01-acc7-ca6b12a65567">
with the boundary condition:
<img width="137" alt="image" src="https://github.com/usnistgov/fipy/assets/155326827/c64ea4fa-a732-4c2e-ab80-e46c557924bf">

I implemented the equation of the first picture, but I want to make sure about setting up the boundary equation. 

`eqI = TransientTerm() == DiffusionTerm(coeff=D)` after setting the coefficient as D,

`X.faceGrad.constrain(bcRight * mesh.faceNormals, mesh.facesRight)` for this line, does bcRight should be J or J/D? Do I have to bring the diffusion coefficient to the RHS?

**Comment by guyer** [comment]
`X.faceGrad` is the gradient of `X`, or $\partial C_A/\partial x$, so constrain it to $J\cdot\hat{n}/D$. Alternatively, [constrain the *flux* at the boundary](https://pages.nist.gov/fipy/en/latest/USAGE.html#applying-fixed-flux-boundary-conditions), in which case, `exteriorFlux` would be $\vec{J}$.

**Reply by beantsja** [reply]
If I use the alternative way, where do I specify the diffusion coefficient D? How does this play in effect if I just let exteriorFlux be J?

**Reply by guyer** [reply]
The coefficient of the `DiffusionTerm` is still $D$. As shown in the link I provided, you can set $D$ to zero at the boundaries where you want the flux to be $J$, but I don't believe it's strictly necessary.

**Reply by beantsja** [reply]
<img width="847" alt="image" src="https://github.com/usnistgov/fipy/assets/155326827/fb283abc-0843-4f79-a564-98df0af99d7c">

In the examples.diffusion.mesh1D document, the boundary condition is:
<img width="131" alt="image" src="https://github.com/usnistgov/fipy/assets/155326827/086f99e9-e10f-4f84-9110-07202696daed">
<img width="796" alt="image" src="https://github.com/usnistgov/fipy/assets/155326827/77518b02-aaca-40a4-ae4e-ec31d621cd20">
But it seems you are putting the diffusion coefficient D on the LHS, so the flux right is just 1, not 1/D. Why is it different here?

**Reply by guyer** [reply]
The example is just wrong and only works because $D = 1$ at the boundaries. About 14 years ago (commit 658a585) we switched from the older `FixedFlux` boundary condition object to the `faceGrad.constrain` mechanism. Unfortunately, we didn't update the discussion and the variable name to reflect that we were changing the boundary condition from a flux constraint to a gradient constraint.

Thanks for bringing this to our attention. I've filed #1000 to fix it.

**Reply by beantsja** [reply]
I set the gradient of X to be J/D as you mentioned. Since this is a concentration gradient, with no time unit involved, I am curious how `eqI = TransientTerm() == DiffusionTerm(coeff=D)` take care of the concentration gradient in each iteration. For flux boundary condition, I think it makes sense that certain amount concentration flows at each time step, but how does the algorithm take care of the concentration gradient?

Another question is:

`mesh = Grid1D(nx=nx, dx=dx)` when I set the initial Grid, if I set nx=10, dx=1 and look at the growth of the concentration at the boundary, the concentration gradient seems to be J/D. However, if the fraction of nx/dx decreases either by decreasing nx or increasing dx, the concentration gradient seems to increase. For example, if nx= 0.01 and dx=10, then, the concentration gradient becomes (J/D)*1000. This seems opposite of what should happen because  when each spatial step length decreases, the flux(or the concentration gradient) also should be adjusted to lower value (My left boundary condition is zero flux).

**Reply by guyer** [reply]
> Since this is a concentration gradient, with no time unit involved, I am curious how `eqI = TransientTerm() == DiffusionTerm(coeff=D)` take care of the concentration gradient in each iteration. For flux boundary condition, I think it makes sense that certain amount concentration flows at each time step, but how does the algorithm take care of the concentration gradient?

`eqI = TransientTerm() == DiffusionTerm(coeff=D)` produces a solution to the concentration as a function of position. Any field that's a function of position has a gradient. In this case, that gradient is constrained at the boundary.

> `mesh = Grid1D(nx=nx, dx=dx)` when I set the initial Grid, if I set nx=10, dx=1 and look at the growth of the concentration at the boundary, the concentration gradient seems to be J/D. However, if the fraction of nx/dx decreases either by decreasing nx or increasing dx, the concentration gradient seems to increase. For example, if nx= 0.01 and dx=10, then, the concentration gradient becomes (J/D)*1000. This seems opposite of what should happen because when each spatial step length decreases, the flux(or the concentration gradient) also should be adjusted to lower value (My left boundary condition is zero flux).

nx= 0.01 and dx=10 doesn't make any sense. nx is the cardinal number of subdivisions in the mesh. [Trying to set nx to anything other than a positive integer should generate an error](https://github.com/usnistgov/fipy/issues/1002).

Leaving `nx=10` and increasing `dx` to 10 leaves the gradient at 1 for me. If that's not what you see, please show explicitly what you're doing.

**Reply by beantsja** [reply]
```python
from fipy import CellVariable, Grid1D, TransientTerm, DiffusionTerm
import math


da = 100  # 1D length. 
nx = 10  # Spatial Grid Points.
dx = da / nx  # Each Spatial Step Length in Domain.
initial_X = 1
D = 1e-6
timeStepDuration = 1
step = 10
X_values = []


mesh = Grid1D(nx=nx, dx=dx)  # Creating the Grid in nx points with dx step length.
X = CellVariable(name="X", mesh=mesh, value=float(initial_X))  # Initializing RH concentration at initial_RH for each grids.
eqI = TransientTerm() == DiffusionTerm(coeff=D)
bcRight = 1
bcLeft = 0

X.faceGrad.constrain(-bcRight* mesh.faceNormals, mesh.facesRight)
X.faceGrad.constrain(bcLeft * mesh.faceNormals, mesh.facesLeft)

for i in range(step):
    eqI.solve(var=X, dt=timeStepDuration)
    X_values.append(X[-1])
    print(X[-1])
    

print(dx)
```

From this code, if I I have da = 100, nx = 10, which gives dx = 10, X[-1] value at the 0th iteration - X[-1] value at the 1st iteration gives 1e-7.  If I have da = 10, nx = 10, which gives dx = 1, the change is 1e-6. If da =1, nx=10, and dx=0.1, then the change is 1e-5. As the dx value decreases, the concentration change is increasing. 

What I am basically trying to do is to set the 1D spatial domain as da = 0.001  to consider 1mm length. But the concentration change gets drastic because with the nx fixed, dx is decreasing as well. The concentration gradient seems to change based on the dx value but not quite sure how it works.

**Reply by guyer** [reply]
There are (at least) two different things at work.
- For dx=1, the location of `X[-1]` is 0.5 units from the right. For dx=0.1, the location of `X[-1]` is 0.05 units from the right. It's not he same location, so there's no reason it should have the same value.
- Probably more significantly, the diffusion time $\tau$ for the domain is $\sqrt{da^2 / D}$. For `da=100`, $\tau=10^{5}$, whereas for `da=10`, $\tau=10^{4}$. In a timestep of 1, the diffusion wave does not propagate as far across the larger domain, so it results in a smaller change in value.

**Reply by beantsja** [reply]
Ok, so the diffusion wave easily propagates for smaller domain.

How should I do in this case?

The concentration gradient is mol/cm^2, and my spatial domain da = 0.1 (cm) with nx = 10. Based on what you are saying, with small diffusion time, the result will change significantly. However, we are also looking at the spatial domain da = 0.1, so the molar concentration must be less compared to da = 1. For example, if there are 10 moles in 1 cm, there should be 1 mole in 0.1 cm. Based on what you say, it seems like there will be more flow of moles in 0.1cm spatial domain than 1cm, and it shows in my coding as well. 

If I want to see the concentration change with small spatial domain, is there something I should be careful about? For example, change the magnitude of the concentration gradient adjusted by the 1cm vs 0.1cm unit?

**Reply by guyer** [reply]
I'm not really understanding what the issue is. The diffusion equation is unconditionally stable, so run whatever time step captures what you're interested in. For `da=0.001`, the diffusion time is 1, so a time step of 1 involves matter diffusing all the way across the domain in one step. If that's not what you want, make the time step smaller. The number of moles in the domain doesn't change any of this.

**Reply by beantsja** [reply]
`from fipy import CellVariable, Grid1D, TransientTerm, DiffusionTerm, Viewer
import math
import matplotlib.pyplot as plt



da = 5e-5   # 1D length. 
nx = 100 # Spatial Grid Points.
dx = da / nx  # Each Spatial Step Length in Domain.
initial_X = ((1)*(5e-5))/(nx)
D = 1e-7 
timeStepDuration = 0.00001
step = 10 #timeStepDuration * second
X_values = []
exteriorFlux = 1e-7
bcLeft = 0
mesh = Grid1D(nx=nx, dx=dx)  # Creating the Grid in nx points with dx step length.
X = CellVariable(name="X", mesh=mesh, value=float(initial_X))  


D.constrain(0., mesh.exteriorFaces)
X.faceGrad.constrain(bcLeft * mesh.faceNormals, mesh.facesLeft)


eqI = (TransientTerm() == DiffusionTerm(coeff=D) + (mesh.exteriorFaces * exteriorFlux).divergence)


for i in range(step):
    eqI.solve(var=X, dt=timeStepDuration)
    X_values.append(float(X[-1]))`
    
 Here I am trying to implementing the situation in my original problem. Instead of using the FaceGrad, I am trying to use the alternative way from your first reply (exterior.Flux). "D.constrain(0., mesh.exteriorFaces)" this line is giving me an error with AttributeError: 'float' object has no attribute 'constrain'. 
  
What would be the issue?

**Reply by guyer** [reply]
`D` must be a `FaceVariable` in order to constrain its values at `mesh.exteriorFaces`.

```python
D = FaceVariable(mesh=mesh, name="D", value=1e-7)
D.constrain(0., mesh.exteriorFaces)
```
```

**Reply by beantsja** [reply]
```python
from fipy import CellVariable, Grid1D, TransientTerm, DiffusionTerm, Viewer
import math

da = 5e-5   # 1D length. cm
nx = 25 # Spatial Grid Points. unitless
dx = da / nx  # Each Spatial Step Length in Domain. cm
initial_X = (2e-6)#((1)*(5e-5))/(nx) #mole
D = 1e-7 
timeStepDuration = 1
step = 100 #timeStepDuration * second
flow = 1e-6
A = 5
n= 2
F =96485
X_values = []


mesh = Grid1D(nx=nx, dx=dx)  # Creating the Grid in nx points with dx step length.
X = CellVariable(name="X", mesh=mesh, value=float(initial_X))  
eqI = TransientTerm() == DiffusionTerm(coeff=D)
bcRight = -((flow*A)) / (n * F*D)
flux = bcRight*D
bcLeft = 0

Xi=sum(X)

X.faceGrad.constrain(bcRight* mesh.faceNormals, mesh.facesRight)
X.faceGrad.constrain(bcLeft * mesh.faceNormals, mesh.facesLeft)

for i in range(step):
    eqI.solve(var=X, dt=timeStepDuration)
    X_values.append(float(X[-1]))
    if sum(X)<0:
        # print(i)
        break

Xf=sum(X)
print("flux:", flux)
print("Total mole:",Xi)
print((Xi-Xf)) #Total mol Change
print(flux*(i+1)) #Total Flux Amount
print((Xi-Xf)*(dx)) #Unit Conversion

```

Here, I am trying to check the total concentration change is equal to the total flux out. I expect `print((Xi-Xf)) #Total mol Change` to be equal to `print(flux*(i+1)) #Total Flux Amount`. However, it is only equal when I do this unit conversion. `print((Xi-Xf)*(dx)) #Unit Conversion`. I am confused with how Fipy takes the input of the initial variables at `X = CellVariable(name="X", mesh=mesh, value=float(initial_X))`. There must be something I am misunderstanding how Fipy is taking control of the variables, and I believe it happens when I locate the initial variable values in each index when mesh is created. But, I am not quite sure what is going on, and why the unit conversion above is necessary.

**Reply by guyer** [reply, marked as answer]
The amount of "stuff" in a cell is the value at the cell center times the volume of the cell. You need to compare the before and after amount of substance (extensive) rather than the concentration (intensive). The preferred way to do that in FiPy (which accounts for nonuniform meshes) is
```python
Xi = (X.cellVolumeAverage * mesh.cellVolumes.sum()).value
  :
  :
Xf = (X.cellVolumeAverage * mesh.cellVolumes.sum()).value
```

`X.cellVolumeAverage` represents $\sum_\text{cells} (X V_\text{cell}) / \sum_\text{cells} V_\text{cell}$.

**note:** if you do not use `.value`, then FiPy's lazy evaluation will result in `Xi` having the same value as `Xf`. You need `.value` to hold the value at that time.

**Reply by beantsja** [reply]
Thank you. The volume of the cell is (the length of the index)^3. Right?

**Reply by guyer** [reply]
In 1D, the volume of the cell is ${\tt dx} \times 1 \times 1$

**Reply by beantsja** [reply]
When I print X, the whole grids, it shows the X values in each index. However, what matters is the X times volume in Fipy, right? Is there a way to print X, but present X*volume values instead of just X? It is because I want to see the results of the actual amount of stuffs in the cell, rather than the value X. 

Another thing I want to confirm is that X*volume is the value that Fipy is using to solve the diffusion equation. Is that correct?

**Reply by guyer** [reply]
`X * mesh.cellVolumes`

> Another thing I want to confirm is that X*volume is the value that Fipy is using to solve the diffusion equation. Is that correct?

Yes. Please see the [documentation on the discretizations that FiPy uses](https://pages.nist.gov/fipy/en/latest/numerical/discret.html#discretization).
